### ISSUE TITLE
Lot's of redundant LDAP queries of listing all groups with members

### ISSUE THREAD

**Issue by rhafer** [owncloud/ocis#4278]
## Describe the bug

A graph request to list all groups with member expanded, becomes very expensive quickly:

`https://localhost:9200/graph/v1.0/groups?$expand=members`

This is partially caused by the backend doing no caching at all, which e.g. means that it is resolving every user that is a member of multiple groups multiple times :cry: (we should at least avoid that to happen within a single graph request)

But the full truth is that this will always  be an expensive query. I think we should enforce a limit of results we return from such a query and introduce support for filters for the graph/groups endpoint. (From a larger numbers of groups and users, listing all of them at once does not provide much value anyway)

**Comment by jvillafanez** [comment]
The easiest option would be to add caching so we don't need to send the same query multiple times knowing the result. We could also include some kind of workers in order to send multiple requests at the same time asynchronously, which could speed up things.

Limiting the number of results could also be a good idea if it's limited in the ldap server, but we'll likely need to deal with "size limit exceeded" errors. It shouldn't be a big deal though.

Note that limiting the results to provide pagination in the sense of offset + limit is a bad idea. The solution for this will be complex and will require proper design. The main problem is that we'll have to deal with the pagination and pagination cookies ourselves, and we need to keep the connection open for a while in case the next page is requested. This means that each ldap request needs to be performed by a worker thread, which will be the one to keep the connection opened (at least for a while).
In addition, this will assume that the pagination will be sequential: first the first page, next the second page, then the third one, and so on. Requesting the fifth page from the beginning will likely force us to request and ignore all the previous pages, and we'll need to start from the beginning if an earlier page is requested afterwards.
Requiring the pagination to always move forward might be too limiting, and taking into account the effort we need to implement that feature correctly it might not be worthy.

-----

As alternative, we could search the users that belong to a group. The expected filter could be something like `(&(objectclass=user)(memberOf=cn=group1,ou=groups,dc=example,dc=com))`. Note that direct access to the user dn should be extremely fast and might be recommended if there are few users in the group; 3-4 dn requests might be faster than having to perform a large search.

In any case, there are several things to notice with this approach:
* Not all the ldap providers have support for the mermberOf filter (openldap requires additional configuration)
* Referential integrity between the group member and the user memberOf attributes might not be guaranteed (openldap requires additional configuration)
* Nested groups might be a problem. Recursive memberOf search might not be available (recursive option is only available in AD I think)

There might be additional cases to take into account since "uniquemember" and "memberuid" are also valid attributes that could be used instead of the usual "member". It might also affect how the memberOf works.

**Comment by jvillafanez** [comment]
I haven't seen code dealing with range requests, and I think we'll also have to include it at some point. This is mainly to deal with large AD groups because there are limits to the number of values that AD will return for an attribute. I think the default limit is 1500, which means that groups with more than 1500 members will need to deal with ranges in order to retrieve the whole list. Linking to https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/e27b48db-6f82-44cd-9038-2e54f790cc1f

**Comment by rhafer** [comment]
> The easiest option would be to add caching so we don't need to send the same query multiple times knowing the result.

Yes. For the `https://localhost:9200/graph/v1.0/groups?$expand=members` query it would already help a lot if we just add a simple DN to LDAP Entry cache that would just live for the lifetime of the request.

> We could also include some kind of workers in order to send multiple requests at the same time asynchronously, which could speed up things.

There were some efforts to add async search support to go-ldap but that stuff is progressing really slow.

> Limiting the number of results could also be a good idea if it's limited in the ldap server, but we'll likely need to deal with "size limit exceeded" errors. It shouldn't be a big deal though.

Yeah. We need to be able to deal with that type of error in any case, as servers usually implement hard limits on the number of
returned entries. 
 
> Note that limiting the results to provide pagination in the sense of offset + limit is a bad idea.

Yes, I don't think that would work.  
 
> As alternative, we could search the users that belong to a group. The expected filter could be something like `(&(objectclass=user)(memberOf=cn=group1,ou=groups,dc=example,dc=com))`.
> Note that direct access to the user dn should be extremely fast and might be recommended if there are few users in the group; 3-4 dn requests might be faster than having to perform a large search.
> 
> In any case, there are several things to notice with this approach:
> 
>     * Not all the ldap providers have support for the mermberOf filter (openldap requires additional configuration)

True. And it seems OpenLDAP has changed the required configuration for that recently. (the memberof overlay was dropped in the latest release AFAIK)

>     * Referential integrity between the group member and the user memberOf attributes might not be guaranteed (openldap requires additional configuration)
> 
>     * Nested groups might be a problem. Recursive memberOf search might not be available (recursive option is only available in AD I think)

Yeah. Don't get me started on nested groups.... ;-)

> There might be additional cases to take into account since "uniquemember" and "memberuid" are also valid attributes that could be used instead of the usual "member".

The current implementation in graph should already support uniquemember and memberuid is just a matter of configuration.

> I haven't seen code dealing with range requests, and I think we'll also have to include it at some point.

Yes, we kind of ignored all AD specifc quirks for now. At some point we have to deal with that.

**Comment by jvillafanez** [comment]
Caching could be done at service level, with an id / name to ldap entry. As long as we have a button to clear the cache easily, we can use https://github.com/owncloud/ocis/pull/4337 so we can keep the information longer (I don't think the information changes frequently) and also could be shared with other graph replicas if needed.

Specific optimizations to providers should be hidden behind the cs3 APIs. For the case of our builtin ldap server, we can optimize the code as much as our provider allows us
